app/models/deck.rb

```diff
@@ -7,4 +7,31 @@ class Deck < ActiveRecord::Base
   validates :title, presence: true, uniqueness: {
               scope: :user_id
             }
+
+  def add_cards(cards)
+    deck_cards = cards.map do |name, quantity|
+      card = Card.find_by(name: name)
+      deck_card = DeckCard.new(card_id: card.id, quantity: quantity)
+    end
+    self.deck_cards = self.deck_cards + deck_cards
+  end
+
+  def sample_deck
+    self.cards.sample(7)
+  end
+
+  def card_type_curve
+    type_count = Hash.new
+    card_type = self.cards.card_type.to_a
+    card_type.each do |type|
+      if type_count.has_key?(type)
+        type_count[type]+= 1
```

Comment:
This is mostly correct but doesn't correctly handle the case where the card quantity is something other than 1. You would need to iterate over the deck cards to account for the quantity correctly. Something like ...

```
def card_type_stats
  stats = {}
  self.deck_cards.include(:cards).each do |deck_card|
    type = deck_card.card.card_type
    if stats.has_key?(type)
      stats[type] += deck_card.quantity
    else
      stats[type] = deck_card.quantity
    end
  end
end
```